crowbar_framework/lib/crowbar/repository.rb

```diff
@@ -0,0 +1,159 @@
+#
+# Copyright 2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Repository
+    attr_reader :platform, :id, :registry
+
+    class << self
+      def load!
+        @config = YAML.load_file(Rails.root.join("config/repos.yml"))
+      end
+
+      def registry
+        load! unless defined? @config
+        @config
+      end
+
+      def where(options = {})
+        platform = options.fetch :platform, nil
+        repo = options.fetch :repo, nil
+        check_all_repos.select do |r|
+          if platform
+            if repo
+              r.platform == platform && r.registry["name"] == repo
+            else
+              r.platform == platform
+            end
+          else
+            r.registry["name"] == repo
+          end
+        end
+      end
+
+      def all_platforms
+        registry.keys
+      end
+
+      def check_all_repos
+        repochecks = []
+        all_platforms.each do |platform|
+          repositories(platform).each do |repo|
+            check = self.new(platform, repo)
+            if platform != Crowbar::Product.ses_platform && Crowbar::Product.is_ses?
+              #FIXME skip the repo check until SES is ported to SLE12-SP1
+              check.registry["required"] = "optional"
+            end
+            repochecks << check
+          end
+        end
+        repochecks
+      end
+
+      def admin_platform
+        NodeObject.admin_node.target_platform
+      end
+
+      def web_port
+        Proposal.where(barclamp: "provisioner").first.raw_attributes["web_port"]
+      end
+
+      def repositories(platform)
+        registry[platform]["repos"].keys
+      end
+
+      def admin_ip
+        NodeObject.admin_node.ip
+      end
+    end
+
+    def initialize(platform, repo)
+      @platform = platform
+      @id = repo
+      @registry = Repository.registry[@platform]["repos"][@id]
+      @url = url
+    end
+
+    def available?
+      all_repo_dirs.include?(@registry["name"]) && check_repo_tag
```

Comment:
I must admit I would expect this method to also check the key to sign the repo. I understand it wasn't done that way because of the view, but I think that's still wrong.

We should have:

-  `available?` that does all the checks
-  `exists?` which check if the repo exists (directory is there, or URL is not empty)
-  `valid_key_file?` which does what you do right now
-  `valid_repo?` which check the repo tag

Then `available?` is really what matters in general, and the view can have the details it needs with the other methods.